flocker/control/_persistence.py

```diff
@@ -21,16 +23,151 @@
+            % (current_version, upgrade_version)
+        )
+        try:
+            migration = getattr(ConfigurationMigration, migration_method)
+        except AttributeError:
+            message = (
+                u"Unable to find a migration path for a version " +
+                unicode(source_version) + u" to version " +
+                unicode(target_version) + u" configuration. " +
+                u"No migration method exists for v" +
+                unicode(current_version) + u" to v" +
+                unicode(upgrade_version) + u"."
+            )
+            raise ConfigurationMigrationError(message)
+        upgraded_config = migration(config)
+        current_version = current_version + 1
+    return upgraded_config
+
+
+class IConfiguration(Interface):
+    """
+    An ``IConfiguration`` implementation provides a serializer and
+    deserializer for a ``Configuration`` model.
+    """
+    def serialize(config):
+        """
+        Serialize the supplied configuration model to JSON.
+
+        :param Configuration config: The configuration to serialize.
+        :return bytes: The JSON representation.
+        """
+
+    def deserialize(config):
+        """
+        Deserialize the supplied JSON to a ``Configuration`` model.
+
+        :param bytes config: The JSON configuration to deserialize.
+        :return Configuration: The configuration model.
+        """
+
 
-class _ConfigurationEncoder(JSONEncoder):
+@implementer(IConfiguration)
+class Configuration_V1(object):
+    """
+    A version 1 configuration.
+    """
+    @classmethod
+    def serialize(cls, config):
+        # Serialized v1 configs are represented as Deployment objects,
+        # not Configuration objects. This is to retain backwards
+        # compatibility.
+        return wire_encode(config.deployment)
+
+    @classmethod
+    def deserialize(cls, config):
+        return Configuration(
+            version=1,
+            deployment=wire_decode(config)
+        )
+
+
+@implementer(IConfiguration)
+class Configuration_V2(object):
+    """
+    A version 2 configuration.
+    """
+    @classmethod
+    def serialize(cls, config):
+        return wire_encode(config)
+
+    @classmethod
+    def deserialize(cls, config):
+        return wire_decode(config)
+
+
+class ConfigurationMigration(object):
+    """
+    Migrate a JSON configuration from one version to another.
+    """
+    @classmethod
+    def configuration_v1_v2(cls, config):
+        """
+        Migrate a v1 JSON configuration to v2.
+
+        :param bytes config: The v1 JSON data.
+        :return bytes: The v2 JSON data.
+        """
+        # In reality, this will transform JSON in some way to ensure
+        # the models line up, e.g. by creating a Leases object,
+        # wire_encode'ing it, then adding that JSON in to the Deployment
+        # object's JSON before deserializing and reserialzing.
+        v1_config = Configuration_V1.deserialize(config)
```

Comment:
This is not a JSON structure transformation, and shouldn't rely on Configuration_V1 AFAICT.

I think given this PR actually introduces a new version of the config you don't need to refer to leases at all. Just try to make this code realistic within the context of this particular upgrade.